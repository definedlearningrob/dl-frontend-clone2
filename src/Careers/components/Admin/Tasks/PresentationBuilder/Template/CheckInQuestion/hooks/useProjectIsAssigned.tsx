import { useMemo } from 'react';
import { useParams } from 'react-router-dom';
import { useQuery } from '@apollo/client';
import { isEmpty } from 'lodash-es';

import teamProjectQuery from '@pbl/graphql/student/queries/teamProject';
import projectQuery from '@pbl/graphql/student/queries/project';

import { useRole } from '@shared/hooks/useRole';

export const useProjectIsAssigned = () => {
  const { isStudent } = useRole();
  const { projectId, teamId } = useParams<{ projectId: string; teamId: string }>();

  const properQuery = useMemo(() => (teamId ? teamProjectQuery : projectQuery), [teamId]);

  const queryVariables = useMemo(
    () =>
      teamId
        ? { id: projectId, teamId, track: true, trackPresentation: true }
        : { id: projectId, track: true, trackPresentation: true },
    [teamId]
  );

  const { data, error, loading } = useQuery(properQuery, {
    variables: queryVariables,
    skip: !isStudent,
  });

  if (!isStudent || loading || error) {
    return false;
  }

  return !isEmpty(data?.project?.assignedAt);
};
