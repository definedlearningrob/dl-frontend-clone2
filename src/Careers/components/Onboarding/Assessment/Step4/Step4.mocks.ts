import { SURVEY_QUESTIONS_QUERY } from '@dc/graphql/student/queries/surveyQuestions';
import CREATE_CAREER_REVIEW_SURVEY_ANSWERS from '@dc/graphql/student/mutations/createCareerReviewSurveyAnswers';
import assessmentAttemptStatusQuery from '@dc/graphql/student/queries/assessmentAttemptStatus';
import finishAssessmentMutation from '@dc/graphql/student/mutations/finishAssessment';

export const createAnswer1Spy = jest.fn();
export const createAnswer2Spy = jest.fn();
export const createAnswer3Spy = jest.fn();
export const createAnswer4Spy = jest.fn();
export const createAnswer5Spy = jest.fn();
export const finishAssessmentSpy = jest.fn();

const createAnswerResponse = {
  data: {
    createCareerReviewSurveyAnswers: {
      status: 'ok',
    },
  },
};

const assessmentProgress = {
  request: {
    query: assessmentAttemptStatusQuery,
  },
  result: () => ({
    data: {
      assessmentProgress: {
        attempt: null,
      },
    },
  }),
};

const finishAssessment = {
  request: {
    query: finishAssessmentMutation,
    variables: { attemptId: '138189' },
  },
  result: () => {
    finishAssessmentSpy();

    return {
      data: {
        createAssessmentResult: {
          status: 'test status',
        },
      },
    };
  },
};

export const step4Mocks = [
  assessmentProgress,
  assessmentProgress,
  assessmentProgress,
  assessmentProgress,
  assessmentProgress,
  finishAssessment,
  {
    request: {
      query: CREATE_CAREER_REVIEW_SURVEY_ANSWERS,
      variables: {
        input: {
          contextId: '138189',
          contextType: 'ASSESSMENT',
          answers: [{ questionId: '7', answer: ['I have no idea'] }],
        },
      },
    },
    result: () => {
      createAnswer1Spy();

      return createAnswerResponse;
    },
  },
  {
    request: {
      query: CREATE_CAREER_REVIEW_SURVEY_ANSWERS,
      variables: {
        input: {
          contextId: '138189',
          contextType: 'ASSESSMENT',
          answers: [
            {
              questionId: '8',
              answer: [
                'Subjects I like or dislike in school',
                'My interests (things that I enjoy)',
                'Test answer',
              ],
            },
          ],
        },
      },
    },
    result: () => {
      createAnswer2Spy();

      return createAnswerResponse;
    },
  },
  {
    request: {
      query: CREATE_CAREER_REVIEW_SURVEY_ANSWERS,
      variables: {
        input: {
          contextId: '138189',
          contextType: 'ASSESSMENT',
          answers: [{ questionId: '9', answer: ['I have no idea what career I want to pursue'] }],
        },
      },
    },
    result: () => {
      createAnswer3Spy();

      return createAnswerResponse;
    },
  },
  {
    request: {
      query: CREATE_CAREER_REVIEW_SURVEY_ANSWERS,
      variables: {
        input: {
          contextId: '138189',
          contextType: 'ASSESSMENT',
          answers: [
            {
              questionId: '10',
              answer: ['They have helped me to decide courses I should consider taking in school'],
            },
          ],
        },
      },
    },
    result: () => {
      createAnswer4Spy();

      return createAnswerResponse;
    },
  },
  {
    request: {
      query: CREATE_CAREER_REVIEW_SURVEY_ANSWERS,
      variables: {
        input: {
          contextId: '138189',
          contextType: 'ASSESSMENT',
          answers: [
            { questionId: '11', answer: ["I'm interested in 3 or more options. (Early Planning)"] },
          ],
        },
      },
    },
    result: () => {
      createAnswer5Spy();

      return createAnswerResponse;
    },
  },
  {
    request: {
      query: SURVEY_QUESTIONS_QUERY,
      variables: {},
    },
    result: {
      data: {
        surveyQuestions: [
          {
            id: '7',
            question: 'Do you know what career you want to pursue?',
            type: 'single_choice',
            options: [
              {
                step: 1,
                option: 'I have no idea',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 2,
                option: "I've thought about it, but I'm not sure",
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 3,
                option: 'I have a general idea',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 4,
                option: 'I know the career I want to pursue',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
            ],
            __typename: 'CareerReviewSurveyQuestion',
          },
          {
            id: '8',
            question:
              'Which of the following personal preferences or traits matter to you when you think about choosing a career?',
            type: 'multiple_choice',
            options: [
              {
                step: 1,
                option: "I haven't thought about choosing a career",
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 2,
                option: 'Subjects I like or dislike in school',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 3,
                option: 'My interests (things that I enjoy)',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 4,
                option: 'My values (things that are important to me)',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 5,
                option: 'My abilities (things I do well)',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 6,
                option: 'Other',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
            ],
            __typename: 'CareerReviewSurveyQuestion',
          },
          {
            id: '9',
            question: 'What do you know about the career(s) you are considering?',
            type: 'multiple_choice',
            options: [
              {
                step: 1,
                option: 'I have no idea what career I want to pursue',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 2,
                option: 'I know what a person does in the career(s)',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 3,
                option: 'I know what courses I need to take in high school',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 4,
                option: 'I know what I need to study in college or trade school',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 5,
                option: 'I know how much money I could make',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 6,
                option: 'I know how easy or hard it is to get a job in the field',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 7,
                option: 'I know what I need to do to enter this career',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 8,
                option: 'Other',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
            ],
            __typename: 'CareerReviewSurveyQuestion',
          },
          {
            id: '10',
            question: 'Which of the following have career-related projects helped you to do?',
            type: 'multiple_choice',
            options: [
              {
                step: 1,
                option: "I've never done a project related to a career",
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 2,
                option: 'They have helped me learn about different career options',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 3,
                option: 'They have helped me learn about what people in specific careers do',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 4,
                option: 'They have helped me learn about which careers I like and dislike',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 5,
                option: 'They have helped me to decide courses I should consider taking in school',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 6,
                option: 'Other',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
            ],
            __typename: 'CareerReviewSurveyQuestion',
          },
          {
            id: '11',
            question: 'How far along are you in your planning for after high school?',
            type: 'single_choice',
            options: [
              {
                step: 1,
                option: "I haven't started thinking about my plan yet. (No Plan)",
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 2,
                option: "I'm interested in 3 or more options. (Early Planning)",
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 3,
                option: "I'm choosing between 2 options. (Narrowed Planning)",
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 4,
                option: 'I have a solid plan for after high school. (Focused Planning)',
                __typename: 'CareerReviewSurveyQuestionOption',
              },
              {
                step: 5,
                option:
                  "I have a plan and I'm already taking steps to make it happen. (Advanced Planning)",
                __typename: 'CareerReviewSurveyQuestionOption',
              },
            ],
            __typename: 'CareerReviewSurveyQuestion',
          },
        ],
      },
    },
  },
];

export const assessmentProgressMock = {
  __typename: 'AssessmentProgress',
  attempt: {
    __typename: 'AssessmentAttempt',
    updatedAt: '2021-09-29T09:00:00Z',
    assessmentType: 'HIGH_SCHOOL',
    id: '138189',
    status: 'IN_PROGRESS',
  },
  interestsAnswers: [],
  status: {
    __typename: 'AssessmentStatus',
    interests: 'COMPLETED',
    studyPreferences: 'COMPLETED',
    workValues: 'COMPLETED',
    reviewSurvey: 'NOT_STARTED',
  },
  studyPreferencesAnswers: [],
  workValuesAnswers: [],
  reviewSurveyAnswers: [],
};
