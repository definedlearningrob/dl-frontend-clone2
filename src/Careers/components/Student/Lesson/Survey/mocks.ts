import lessonInCourseQuery from '@dc/graphql/student/queries/lessonInCourse';
import { CREATE_CAREER_REVIEW_SURVEY_ATTEMPT } from '@dc/graphql/student/mutations/createCareerReviewSurveyAttempt';
import createCareerReviewSurveyAnswersMutation from '@dc/graphql/student/mutations/createCareerReviewSurveyAnswers';
import currentCoursesQuery from '@dc/graphql/student/queries/currentCourses';
import { SURVEY_QUESTION_TYPE } from '@dc/components/Student/Lesson/Survey/types';

export const createAttemptSpy = jest.fn();
export const lessonInCourseSpy = jest.fn();

export let createCareerReviewSurveyAnswersSpy = jest.fn();
export let currentCoursesSpy = jest.fn();

export const mocks = [
  {
    request: {
      query: lessonInCourseQuery,
      variables: { courseId: '2', lessonId: '111', track: true },
    },
    result: () => {
      lessonInCourseSpy();

      return {
        data: {
          course: {
            careerName: 'Dental Laboratory Technicians',
            content: {
              checkIns: [],
              items: [],
            },
            hasInstitutionsInStudentState: true,
            id: '2',
            lesson: {
              assignments: [],
              attachments: [],
              careerReviewSurvey: {
                version: 1,
                performed: false,
                questions: [
                  {
                    answer: [],
                    id: '1',
                    options: [
                      {
                        option: 'Yes, definitely',
                        step: 1,
                      },
                      {
                        option: 'Definitely not',
                        step: 2,
                      },
                    ],
                    question: 'Can blind people see their dreams?',
                    type: 'single_choice' as SURVEY_QUESTION_TYPE,
                  },
                  {
                    answer: [],
                    id: '2',
                    options: [
                      {
                        option: 'Very Interested',
                        step: 1,
                      },
                      {
                        option: 'Not at all Interested',
                        step: 2,
                      },
                      {
                        option: 'Other',
                        step: 3,
                      },
                    ],
                    question: 'How interested are you in the Pathway this course is a part of?',
                    type: 'multiple_choice' as SURVEY_QUESTION_TYPE,
                  },
                ],
                __typename: 'CareerReviewSurvey',
              },
              checkInGroups: [],
              checkInQuestions: [],
              description: {
                introduction: null,
                goal: null,
                role: null,
                audience: null,
                situation: null,
                __typename: 'LessonDescription',
              },
              hasPresentation: false,
              id: '609',
              imageUrl: 'http://test.com/image.png',
              name: 'Career Review Survey',
              externalPresentations: [],
              researchLinks: [],
              texts: [],
              type: 'career_review_survey',
              videos: [],
              vocabularies: [],
              __typename: 'Lesson',
            },
            lessons: [
              {
                id: '782',
                step: 4,
                careerReviewSurvey: null,
              },
              {
                id: '609',
                step: 5,
                careerReviewSurvey: {
                  performed: false,
                  version: 1,
                  __typename: 'CareerReviewSurvey',
                },
              },
              {
                id: '275',
                step: 1,
                careerReviewSurvey: null,
              },
              {
                id: '331',
                step: 2,
                careerReviewSurvey: null,
              },
              {
                id: '668',
                step: 3,
                careerReviewSurvey: null,
              },
            ],
            name: 'Dental Laboratory Technician',
            progress: {
              submitted: 10,
              total: 10,
            },
            reviewSurvey: {
              questions: [
                {
                  id: '1',
                },
                {
                  id: '2',
                },
                {
                  id: '3',
                },
                {
                  id: '4',
                },
                {
                  id: '5',
                },
                {
                  id: '6',
                },
                {
                  id: '7',
                },
                {
                  id: '8',
                },
                {
                  id: '9',
                },
                {
                  id: '10',
                },
                {
                  id: '11',
                },
              ],
            },
          },
        },
      };
    },
  },
  {
    request: {
      query: CREATE_CAREER_REVIEW_SURVEY_ATTEMPT,
      variables: { input: { contextType: 'COURSE', contextId: '1' } },
    },
    result: () => {
      createAttemptSpy();

      return {
        data: {
          createCareerReviewSurveyAttempt: {
            careerReviewSurveyAttempt: {
              id: '2',
              status: 'IN_PROGRESS',
            },
          },
        },
      };
    },
  },
  {
    request: {
      query: createCareerReviewSurveyAnswersMutation,
      variables: {
        input: {
          contextId: '1',
          contextType: 'COURSE',
          answers: [
            { questionId: '1', answer: ['Yes, definitely'] },
            { questionId: '2', answer: ['Very Interested'] },
          ],
        },
      },
    },
    result: () => {
      createCareerReviewSurveyAnswersSpy();

      return {
        data: {
          createCareerReviewSurveyAnswers: {
            status: 'success',
          },
        },
      };
    },
  },
  {
    request: {
      query: currentCoursesQuery,
    },
    result: () => {
      currentCoursesSpy();

      return {
        data: {
          currentCourses: {
            id: '1',
            imageUrl: '',
            name: '',
            progress: {
              submitted: 2,
              total: 5,
            },
            status: 'success',
            pathway: {
              name: '',
            },
            thumbnailUrl: '',
            type: 'survey',
          },
        },
      };
    },
  },
];
