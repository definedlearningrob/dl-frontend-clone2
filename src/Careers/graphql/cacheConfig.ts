import { FieldFunctionOptions, InMemoryCache, StoreObject } from '@apollo/client';
import { relayStylePagination } from '@apollo/client/utilities';

import { dualPagination } from '@shared/utils/cache';

export default new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        badges: dualPagination(),
        units: dualPagination(),
        tags: dualPagination(),
        announcements: relayStylePagination(),
        allCourses: dualPagination(),
        courses: dualPagination(),
        conversationGroups: relayStylePagination(),
        recentApplications: relayStylePagination(),
        entities: dualPagination(),
        extensionFields: {
          merge({ nodes = [] } = {}, incoming) {
            return {
              nodes: [...nodes, ...incoming.nodes],
              pagesCount: incoming.pagesCount,
              nodesCount: incoming.nodesCount,
            };
          },
          read(existing) {
            return existing;
          },
        },
        notifications: {
          merge({ nodes = [] } = {}, incoming) {
            return {
              nodes: [...nodes, ...incoming.nodes],
              pagesCount: incoming.pagesCount,
              nodesCount: incoming.nodesCount,
            };
          },
          read(
            { nodes = [], pagesCount, nodesCount } = {},
            { readField, args }: FieldFunctionOptions
          ) {
            const newNodes = nodes.filter((node: StoreObject) => {
              const nodeType = readField('type', node);

              return args?.type === nodeType;
            });

            return {
              nodes: newNodes,
              pagesCount,
              nodesCount,
            };
          },
        },
        institutions: {
          merge({ nodes = [] } = {}, incoming) {
            return {
              nodes: [...nodes, ...incoming.nodes],
              pagesCount: incoming.pagesCount,
              nodesCount: incoming.nodesCount,
            };
          },
          read(existing) {
            return existing;
          },
        },
        opportunities: {
          merge({ nodes = [] } = {}, incoming, { variables }) {
            return {
              //@ts-ignore
              nodes: variables.infiniteScroll ? [...nodes, ...incoming.nodes] : incoming.nodes,
              pagesCount: incoming.pagesCount,
              nodesCount: incoming.nodesCount,
            };
          },
          read(existing) {
            return existing;
          },
        },
        partners: {
          merge({ nodes = [] } = {}, incoming, { variables }) {
            return {
              //@ts-ignore
              nodes: variables.infiniteScroll ? [...nodes, ...incoming.nodes] : incoming.nodes,
              pagesCount: incoming.pagesCount,
              nodesCount: incoming.nodesCount,
            };
          },
          read(existing) {
            return existing;
          },
        },
      },
    },
    AdminDashboard: {
      keyFields: ['userId'],
      fields: {
        entity: {
          merge: true,
        },
      },
    },
    Conversation: {
      fields: {
        messages: relayStylePagination(),
      },
    },
    ConversationGroup: {
      keyFields: ['participant', ['uuid']],
    },
    Institution: {
      fields: {
        programs: {
          merge({ nodes = [] } = {}, incoming, { variables }) {
            return {
              //@ts-ignore
              nodes: variables.infiniteScroll ? [...nodes, ...incoming.nodes] : incoming.nodes,
              pagesCount: incoming.pagesCount,
              nodesCount: incoming.nodesCount,
            };
          },
          read(existing) {
            return existing;
          },
        },
      },
    },
    Entity: {
      keyFields: ['uuid'],
      merge: true,
    },
    EntityData: {
      keyFields: ['uuid'],
      fields: {
        settings: {
          merge: true,
        },
      },
    },
    EntitySettings: {
      merge: true,
    },
    Portfolio: {
      fields: {
        projects: relayStylePagination(['$type']),
      },
      keyFields: ['studentId'],
    },
    SchoolClass: {
      keyFields: ['uuid'],
      fields: {
        students: {
          merge({ nodes = [] } = {}, incoming, { readField }) {
            const newNodes = nodes.filter(
              (node: StoreObject) =>
                !incoming.nodes.find(
                  (incomingNode: StoreObject) =>
                    readField('uuid', incomingNode) === readField('uuid', node)
                )
            );

            return { nodes: [...newNodes, ...incoming.nodes], pagesCount: incoming.pagesCount };
          },
          read(existing) {
            return existing;
          },
        },
      },
    },
    Opportunity: {
      merge: true,
      fields: {
        virtualInternship: {
          merge: true,
        },
        entities: {
          merge: false,
        },
      },
    },
    Student: {
      keyFields: (data) => {
        const hasUUID = Object.keys(data).some((fieldName) => fieldName === 'uuid');

        if (hasUUID) return ['uuid'];

        return false;
      },
    },
    TeacherDashboard: {
      keyFields: ['userId'],
      fields: {
        activityLog: relayStylePagination(),
      },
    },

    UserInfo: {
      keyFields: ['uuid'],
    },
    CommonAppData: {
      merge: true,
    },
    LessonItem: {
      keyFields: false,
    },
    Tag: {
      merge: true,
    },
    RecommendationRequest: {
      keyFields: ['applicant', ['uuid']],
    },
    PortfolioProject: {
      keyFields: ['id', 'resourceClass'],
    },
    PlanReport: {
      merge: true,
    },
    Reports: {
      merge: false,
    },
    PortfolioSubmission: {
      merge: true,
    },
    Badge: {
      keyFields: (data) => {
        const hasResource = Object.keys(data).some((fieldName) => fieldName === 'resource');

        if (hasResource) return ['id', 'resource'];

        return ['id'];
      },
    },
    StudentBadge: {
      keyFields: ['id', 'resource'],
    },
    Catalog: {
      merge: true,
    },
  },
});
