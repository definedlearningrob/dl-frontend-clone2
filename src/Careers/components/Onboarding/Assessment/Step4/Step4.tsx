import { useState } from 'react';
import { Formik } from 'formik';

import { Step4Content } from '@dc/components/Onboarding/Assessment/Step4/Step4Content/Step4Content';
import { SplashScreen } from '@dc/components/Onboarding/Assessment/SplashScreen/SplashScreen';
import { useSurveyQuestionsQuery } from '@dc/graphql/student/hooks/useSurveyQuestionsQuery';

type SurveyFormType = {
  [questionId: string]: {
    answers: string[];
    otherAnswer: string;
  };
};

type Props = {
  assessmentType: string;
  isFirstQuestion: boolean;
};

export const Step4 = ({ assessmentType, isFirstQuestion }: Props) => {
  const [displaySplashScreen, setDisplaySplashScreen] = useState(isFirstQuestion);
  const { data } = useSurveyQuestionsQuery();

  if (displaySplashScreen)
    return (
      <SplashScreen
        assessmentType={assessmentType}
        setDisplaySplashScreen={setDisplaySplashScreen}
        step={4}
      />
    );

  if (!data) return null;

  const surveyQuestions = data.surveyQuestions;

  const initialValues: SurveyFormType = surveyQuestions.reduce((acc, { id, type }) => {
    const answers = type === 'single_choice' ? '' : [];

    return { ...acc, [id]: { answers, otherAnswer: '' } };
  }, {});

  return (
    <Formik initialValues={initialValues} onSubmit={() => {}}>
      <Step4Content assessmentType={assessmentType} surveyQuestions={surveyQuestions} />
    </Formik>
  );
};
