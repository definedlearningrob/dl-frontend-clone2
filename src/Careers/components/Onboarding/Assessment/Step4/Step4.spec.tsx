import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { MemoryHistory, createMemoryHistory } from 'history';
import { act, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Route } from 'react-router-dom';

import { renderWithRouterAndReduxProvider } from '@dc/utils/test';
import { UserInfoProvider } from '@dc/hooks/useUserInfo';
import { userInfoMock } from '@dc/tests/mocks/userMocks';
import cacheConfig from '@dc/graphql/cacheConfig';
import {
  assessmentProgressMock,
  createAnswer1Spy,
  createAnswer2Spy,
  createAnswer3Spy,
  createAnswer4Spy,
  createAnswer5Spy,
  finishAssessmentSpy,
  step4Mocks,
} from '@dc/components/Onboarding/Assessment/Step4/Step4.mocks';
import { AssessmentProvider } from '@dc/hooks/useAssessment';

import { Step4 } from './Step4';

describe('Step 4 - post-assessment career review survey', () => {
  const renderStep4 = ({
    history,
    mocks = [],
  }: { history?: MemoryHistory; mocks?: MockedResponse[] } = {}) =>
    renderWithRouterAndReduxProvider(
      <MockedProvider cache={cacheConfig} mocks={[...mocks, userInfoMock]}>
        <UserInfoProvider>
          <Route path='/onboarding/assessment/step/:stepNumber/question/:questionNumber'>
            <AssessmentProvider assessment={assessmentProgressMock}>
              <Step4 assessmentType='sas' isFirstQuestion={true} />
            </AssessmentProvider>
          </Route>
        </UserInfoProvider>
      </MockedProvider>,
      { history }
    );

  it('should render info page correctly', () => {
    const history = createMemoryHistory({
      initialEntries: ['/onboarding/assessment/step/4/question/1'],
    });

    const { container } = renderStep4({ history, mocks: step4Mocks });

    const heading = screen.getByRole('heading', {
      name: 'Section 4: Career Review Survey',
      level: 2,
    });

    const beginButton = screen.getByRole('button', { name: 'Begin' });

    expect(heading).toBeInTheDocument();
    expect(beginButton).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should go through whole survey correctly', async () => {
    const history = createMemoryHistory({
      initialEntries: ['/onboarding/assessment/step/4/question/1'],
    });

    renderStep4({ history, mocks: step4Mocks });

    const heading = screen.getByRole('heading', {
      name: 'Section 4: Career Review Survey',
      level: 2,
    });

    const beginButton = screen.getByRole('button', { name: 'Begin' });

    expect(heading).toBeInTheDocument();
    expect(beginButton).toBeInTheDocument();

    userEvent.click(beginButton);

    // QUESTION 1

    expect(
      await screen.findByRole('heading', { name: 'Do you know what career you want to pursue?' })
    ).toBeInTheDocument();

    const question1answer1 = screen.getByRole('radio', { name: 'I have no idea' });
    const question1answer2 = screen.getByRole('radio', {
      name: "I've thought about it, but I'm not sure",
    });
    const question1answer3 = screen.getByRole('radio', { name: 'I have a general idea' });
    const question1answer4 = screen.getByRole('radio', {
      name: 'I know the career I want to pursue',
    });

    expect(question1answer1).toBeInTheDocument();
    expect(question1answer2).toBeInTheDocument();
    expect(question1answer3).toBeInTheDocument();
    expect(question1answer4).toBeInTheDocument();

    await act(async () => {
      userEvent.click(question1answer1);
    });

    const nextButtonAfterQuestion1 = screen.getByRole('button', { name: 'Next Question' });

    userEvent.click(nextButtonAfterQuestion1);

    await waitFor(() => expect(createAnswer1Spy).toHaveBeenCalledTimes(1));

    // QUESTION 2

    expect(
      await screen.findByRole('heading', {
        name: 'Which of the following personal preferences or traits matter to you when you think about choosing a career?',
      })
    ).toBeInTheDocument();

    const question2Answers = [
      screen.getByRole('checkbox', {
        name: "I haven't thought about choosing a career",
      }),
      screen.getByRole('checkbox', {
        name: 'Subjects I like or dislike in school',
      }),
      screen.getByRole('checkbox', {
        name: 'My interests (things that I enjoy)',
      }),
      screen.getByRole('checkbox', {
        name: 'My values (things that are important to me)',
      }),
      screen.getByRole('checkbox', {
        name: 'My abilities (things I do well)',
      }),
      screen.getByRole('checkbox', {
        name: 'Other',
      }),
    ];

    question2Answers.forEach((answer) => {
      expect(answer).toBeInTheDocument();
    });

    userEvent.click(question2Answers[1]);
    userEvent.click(question2Answers[2]);
    userEvent.click(question2Answers[5]);

    const otherAnswerField = screen.getByPlaceholderText('For me the most matter...');

    userEvent.paste(otherAnswerField, 'Test answer');

    const nextButtonAfterQuestion2 = screen.getByRole('button', { name: 'Next Question' });

    userEvent.click(nextButtonAfterQuestion2);
    await waitFor(() => expect(createAnswer2Spy).toHaveBeenCalledTimes(1));

    // QUESTION 3
    expect(
      await screen.findByRole('heading', {
        name: 'What do you know about the career(s) you are considering?',
      })
    ).toBeInTheDocument();

    const question3answers = [
      screen.getByRole('checkbox', {
        name: 'I have no idea what career I want to pursue',
      }),
      screen.getByRole('checkbox', {
        name: 'I know what a person does in the career(s)',
      }),
      screen.getByRole('checkbox', {
        name: 'I know what courses I need to take in high school',
      }),
      screen.getByRole('checkbox', {
        name: 'I know what I need to study in college or trade school',
      }),
      screen.getByRole('checkbox', {
        name: 'I know how much money I could make',
      }),
      screen.getByRole('checkbox', {
        name: 'I know how easy or hard it is to get a job in the field',
      }),
      screen.getByRole('checkbox', {
        name: 'I know what I need to do to enter this career',
      }),
      screen.getByRole('checkbox', {
        name: 'Other',
      }),
    ];

    question3answers.forEach((answer) => {
      expect(answer).toBeInTheDocument();
    });

    await act(async () => {
      userEvent.click(question3answers[0]);
    });

    const disabledAnswers = question3answers.slice(1, 7);

    disabledAnswers.forEach((answer) => {
      expect(answer).toBeDisabled();
    });

    const nextButtonAfterQuestion3 = screen.getByRole('button', { name: 'Next Question' });

    userEvent.click(nextButtonAfterQuestion3);

    await waitFor(() => expect(createAnswer3Spy).toHaveBeenCalledTimes(1));

    //QUESTION 4

    expect(
      await screen.findByRole('heading', {
        name: 'Which of the following have career-related projects helped you to do?',
      })
    ).toBeInTheDocument();

    const question4answers = [
      screen.getByRole('checkbox', {
        name: "I've never done a project related to a career",
      }),
      screen.getByRole('checkbox', {
        name: 'They have helped me learn about different career options',
      }),
      screen.getByRole('checkbox', {
        name: 'They have helped me learn about what people in specific careers do',
      }),
      screen.getByRole('checkbox', {
        name: 'They have helped me learn about which careers I like and dislike',
      }),
      screen.getByRole('checkbox', {
        name: 'They have helped me to decide courses I should consider taking in school',
      }),
      screen.getByRole('checkbox', {
        name: 'Other',
      }),
    ];

    question4answers.forEach((answer) => {
      expect(answer).toBeInTheDocument();
    });

    await act(async () => {
      userEvent.click(question4answers[4]);
    });

    const nextButtonAfterQuestion4 = screen.getByRole('button', { name: 'Next Question' });

    userEvent.click(nextButtonAfterQuestion4);

    await waitFor(() => expect(createAnswer4Spy).toHaveBeenCalledTimes(1));

    // QUESTION 5
    expect(
      await screen.findByRole('heading', {
        name: 'How far along are you in your planning for after high school?',
      })
    ).toBeInTheDocument();

    const question5answers = [
      screen.getByRole('radio', {
        name: "I haven't started thinking about my plan yet. (No Plan)",
      }),
      screen.getByRole('radio', {
        name: "I'm interested in 3 or more options. (Early Planning)",
      }),
      screen.getByRole('radio', {
        name: "I'm choosing between 2 options. (Narrowed Planning)",
      }),
      screen.getByRole('radio', {
        name: 'I have a solid plan for after high school. (Focused Planning)',
      }),
      screen.getByRole('radio', {
        name: "I have a plan and I'm already taking steps to make it happen. (Advanced Planning)",
      }),
    ];

    question5answers.forEach((answer) => {
      expect(answer).toBeInTheDocument();
    });

    await act(async () => {
      userEvent.click(question5answers[1]);
    });

    const nextButtonAfterQuestion5 = screen.getByRole('button', { name: 'Next Question' });

    userEvent.click(nextButtonAfterQuestion5);

    await waitFor(() => expect(createAnswer5Spy).toHaveBeenCalledTimes(1));
    await waitFor(() => expect(finishAssessmentSpy).toHaveBeenCalledTimes(1));
  });
});
