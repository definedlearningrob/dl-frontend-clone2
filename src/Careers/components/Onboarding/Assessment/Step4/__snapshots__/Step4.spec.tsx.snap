// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step 4 - post-assessment career review survey should render info page correctly 1`] = `
<div>
  <section
    class="flex flex-col pt-base xxxl:pt-[150px] items-center"
  >
    <h2
      class="text-font-primary text-2lg mb-sm"
    >
      Section 4: Career Review Survey
    </h2>
    <p
      class="text-center text-base xxxl:w-1/2 mb-lg"
    >
      Goal: Assess your knowledge of career options, opportunities, and planning for the future.
    </p>
    <div
      class="max-w-[1200px] flex"
    >
      <div
        class="splash-screen__player"
        data-testid="splash-video-player"
        style="width: 640px; height: 360px;"
      />
      <ol
        class="splash-screen__instruction-list"
      >
        <li
          class="splash-screen__instruction-item leading-lg text-font-secondary"
        >
          1. In this section, you will be asked to answer 5 questions. The questions you will encounter are displayed in a variety of formats.
        </li>
        <li
          class="splash-screen__instruction-item leading-lg text-font-secondary"
        >
          2. Click the “Next Question” button to proceed. You will complete this process 5 times.
        </li>
        <button
          class="button flex items-center justify-center gap-xs rounded-sm cursor-pointer text-sm font-medium no-underline focus-visible:outline focus-visible:outline-1 focus-visible:outline-offset-[3px] min-w-[200px] w-fit mx-auto -primary bg-primary-500 hover:bg-primary-600 outline-primary-500 border-primary-500 py-xs px-base leading-[22px] border text-white"
          data-testid="button"
          id="splash-screen-button"
          type="button"
        >
          <span
            class="items-center flex justify-center gap-xs"
          >
            Begin
          </span>
        </button>
      </ol>
    </div>
  </section>
</div>
`;
